src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs

```diff
@@ -4,19 +4,72 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
-using Microsoft.Build.BuildCheck.Analyzers;
+using System.Reflection;
+using Microsoft.Build.BackEnd.Logging;
 using Microsoft.Build.BuildCheck.Infrastructure;
+using Microsoft.Build.Experimental.BuildCheck;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
 
 namespace Microsoft.Build.BuildCheck.Acquisition;
 
-internal class BuildCheckAcquisitionModule
+internal class BuildCheckAcquisitionModule : IBuildCheckAcquisitionModule
 {
-    private static T Construct<T>() where T : new() => new();
-    public BuildAnalyzerFactory CreateBuildAnalyzerFactory(AnalyzerAcquisitionData analyzerAcquisitionData)
+    private readonly ILoggingService _loggingService;
+
+    internal BuildCheckAcquisitionModule(ILoggingService loggingService)
+    {
+        _loggingService = loggingService;
+    }
+
+#if FEATURE_ASSEMBLYLOADCONTEXT
+    /// <summary>
+    /// AssemblyContextLoader used to load DLLs outside of msbuild.exe directory
+    /// </summary>
+    private static readonly CoreClrAssemblyLoader s_coreClrAssemblyLoader = new();
+#endif
+
+    /// <summary>
+    /// Creates a list of factory delegates for building analyzer rules instances from a given assembly path.
+    /// </summary>
+    public List<BuildAnalyzerFactory> CreateBuildAnalyzerFactories(AnalyzerAcquisitionData analyzerAcquisitionData, BuildEventContext buildEventContext)
     {
-        // Acquisition module - https://github.com/dotnet/msbuild/issues/9633
-        return Construct<SharedOutputPathAnalyzer>;
+        var analyzersFactories = new List<BuildAnalyzerFactory>();
+
+        try
+        {
+            Assembly? assembly = null;
+#if FEATURE_ASSEMBLYLOADCONTEXT
+            assembly = s_coreClrAssemblyLoader.LoadFromPath(analyzerAcquisitionData.AssemblyPath);
+#else
+            assembly = Assembly.LoadFrom(analyzerAcquisitionData.AssemblyPath);
+#endif
+
+            IEnumerable<Type> analyzerTypes = assembly.GetTypes().Where(t => typeof(BuildAnalyzer).IsAssignableFrom(t));
+
+            foreach (Type analyzerType in analyzerTypes)
+            {
+                if (Activator.CreateInstance(analyzerType) is BuildAnalyzer instance)
+                {
+                    analyzersFactories.Add(() => instance);
```

Comment:
I haven't reviewed everything yet but this looks unexpected. Presumably, the reason why this method returns a list of _factories_ is to allow analyzers to be loaded lazily. If it is desired to load and scan the assembly here, shouldn't the return type change to the list of analyzers?

Reply:
@JanKrivanek this question can be addressed to you :)

Reply:
This is meant as preparation for future (hopefully still V1) improvement, where we do not load the type (nor even scan the assembly) before it is required (we hit a project for which it's enabled by editorconfig - as there is a chance it might not be enabled for current build at all).

To enable this, the analyzer package author would need to put the rule ids and default enablement status as an argument to the intrinsic function (the template can handle this by calling the type and extracting the values and adding it as arguments to the generated intrinsic function call)

This is already supported by the infra: https://github.com/dotnet/msbuild/blob/9cdb3615adb4115f92b390de2f258fac5f320909/src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs#L208-L222

Reply:
Sounds like a reasonable future proofing. Thank you and feel free to resolve.